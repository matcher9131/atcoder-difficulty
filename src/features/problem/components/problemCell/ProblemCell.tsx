import clsx from "clsx";
import type { ReactNode } from "react";

export type ProblemCellProps = {
    readonly iconHref: string;
    readonly fillColor: string;
    readonly difficulty: string;
    readonly displayName: string;
    readonly textColor: string;
    readonly linkHref: string;
};

export const ProblemCell = ({
    fillColor,
    iconHref,
    difficulty,
    displayName,
    textColor,
    linkHref,
}: ProblemCellProps): ReactNode => {
    return (
        <td className="px-3 py-6 gap-x-1">
            <div className="flex items-baseline">
                <span className="relative">
                    <svg role="img" aria-label="難易度アイコン" className={clsx("w-6", "h-4", fillColor)}>
                        <use href={iconHref} />
                    </svg>
                    <div className="absolute top-4 w-6 text-center text-[.5em]">{difficulty}</div>
                </span>
                <a href={linkHref} target="_blank" rel="noreferrer" className={clsx("truncate", textColor)}>
                    {displayName}
                </a>
            </div>
        </td>
    );
};
