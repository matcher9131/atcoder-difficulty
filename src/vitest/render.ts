import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
