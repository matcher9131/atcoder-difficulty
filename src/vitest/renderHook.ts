import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
