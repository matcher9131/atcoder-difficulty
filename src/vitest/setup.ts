import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
