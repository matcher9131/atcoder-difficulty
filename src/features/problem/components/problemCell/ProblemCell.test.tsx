import { describe, it, expect, vi, beforeEach } from "vitest";
import { screen, fireEvent } from "@testing-library/react";
import { ProblemCell, type ProblemCellProps } from "./ProblemCell";
import { render } from "../../../../vitest/render";

// src/features/problem/components/problemCell/ProblemCell.test.tsx

// Mock useTranslation
vi.mock("react-i18next", () => ({
    useTranslation: () => ({
        t: (key: string) => key,
    }),
}));

const defaultProps: ProblemCellProps = {
    iconHref: "#icon-difficulty",
    fillColor: "fill-red-500",
    difficulty: "1234",
    solveProbability: "56%",
    displayName: "Sample Problem",
    textColor: "text-blue-700",
    linkHref: "https://atcoder.jp/problems/abc123_a",
    problemIndex: "A",
    graphIconHref: "#icon-graph",
    graphButtonIsEnabled: true,
    onGraphButtonClick: vi.fn(),
};

describe("ProblemCell", () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    it("renders SVG icon with correct attributes", () => {
        render(<ProblemCell {...defaultProps} />);
        const svg = screen.getByRole("img", { name: "problemCell.difficultyIconLabel" });
        expect(svg).toBeInTheDocument();
        expect(svg).toHaveClass("w-6", "h-4", defaultProps.fillColor);
        expect(svg.querySelector("use")).not.toBeNull();
        expect(svg.querySelector("use")).toHaveAttribute("href", defaultProps.iconHref);
    });

    it("renders difficulty and solveProbability when not empty", () => {
        render(<ProblemCell {...defaultProps} />);
        expect(screen.getByText(defaultProps.difficulty)).toBeInTheDocument();
        expect(screen.getByText(defaultProps.solveProbability)).toBeInTheDocument();
    });

    it("renders displayName with problemIndex prefix", () => {
        render(<ProblemCell {...defaultProps} />);
        if (defaultProps.problemIndex == null) throw new Error("problemIndex is required");
        expect(screen.getByText(`${defaultProps.problemIndex} - ${defaultProps.displayName}`)).toBeInTheDocument();
    });

    it("renders displayName without problemIndex prefix", () => {
        render(<ProblemCell {...defaultProps} problemIndex={undefined} />);
        expect(screen.getByText(defaultProps.displayName)).toBeInTheDocument();
    });

    it("renders link with correct attributes and classes", () => {
        render(<ProblemCell {...defaultProps} />);
        const link = screen.getByRole("link");
        expect(link).toHaveAttribute("href", defaultProps.linkHref);
        expect(link).toHaveAttribute("target", "_blank");
        expect(link).toHaveAttribute("rel", "noreferrer");
        expect(link).toHaveClass(
            "w-full",
            "text-left",
            "truncate",
            defaultProps.textColor,
            "link",
            "link-hover",
            "hover:brightness-75",
        );
    });

    it("renders graph button when enabled", () => {
        render(<ProblemCell {...defaultProps} />);
        const button = screen.getByRole("button");
        expect(button).toBeInTheDocument();
        const svg = button.querySelector("svg");
        expect(svg).toHaveAttribute("role", "img");
        expect(svg).toHaveAttribute("aria-label", "Show graph");
        expect(svg?.querySelector("use")).toHaveAttribute("href", defaultProps.graphIconHref);
        expect(svg?.querySelector("title")).not.toBeNull();
        expect(svg?.querySelector("title")?.textContent).toBe("problemCell.graphButtonIconTitle");
    });

    it("does not render graph button when disabled", () => {
        render(<ProblemCell {...defaultProps} graphButtonIsEnabled={false} />);
        expect(screen.queryByRole("button")).not.toBeInTheDocument();
    });

    it("calls onGraphButtonClick when graph button is clicked", () => {
        render(<ProblemCell {...defaultProps} />);
        const button = screen.getByRole("button");
        fireEvent.click(button);
        expect(defaultProps.onGraphButtonClick).toHaveBeenCalledTimes(1);
    });

    it("sets tabIndex and accessibility attributes", () => {
        render(<ProblemCell {...defaultProps} />);
        const td = screen.getByRole("cell");
        expect(td).toHaveAttribute("tabIndex", "0");
    });

    it("matches snapshot", () => {
        const { container } = render(<ProblemCell {...defaultProps} />);
        expect(container).toMatchSnapshot();
    });
});
