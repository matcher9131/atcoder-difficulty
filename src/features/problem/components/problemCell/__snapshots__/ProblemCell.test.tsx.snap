// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProblemCell > matches snapshot 1`] = `
<div>
  <td
    class="flex px-3 py-auto h-18 items-center border-t-1 border-base-content/75 group"
    tabindex="0"
  >
    <span
      class="relative"
    >
      <svg
        aria-label="problemCell.difficultyIconLabel"
        class="w-6 h-4 fill-red-500"
        role="img"
      >
        <use
          href="#icon-difficulty"
        />
      </svg>
      <div
        class="absolute -top-4 w-6 text-center text-[.7em]"
      >
        1234
      </div>
      <div
        class="absolute top-4 w-6 text-center text-[.7em]"
      >
        56%
      </div>
    </span>
    <a
      class="w-full text-left truncate text-blue-700 link link-hover hover:brightness-75"
      href="https://atcoder.jp/problems/abc123_a"
      rel="noreferrer"
      target="_blank"
    >
      A - Sample Problem
    </a>
    <button
      class="hidden group-hover:block group-focus-within:block"
    >
      <svg
        aria-label="Show graph"
        class="h-6 w-6 fill-current cursor-pointer hover:invert-30 transition-all"
        role="img"
      >
        <title>
          problemCell.graphButtonIconTitle
        </title>
        <use
          href="#icon-graph"
        />
      </svg>
    </button>
  </td>
</div>
`;
